### PR TITLE
ci: add build app e2e test for TypeScript

### PR DESCRIPTION
**Pull request by dpilch**
#### Description of changes

A recent change (https://github.com/aws-amplify/amplify-codegen/pull/489) caused invalid model generation in some cases. This PR introduces E2E tests that attempt to build an app with the generated models. This is starting with a basic set of test using the [schema test library](https://github.com/aws-amplify/amplify-category-api/blob/main/packages/amplify-graphql-schema-test-library/src/index.ts) so that coverage for the error is introduced sooner rather than later.

Future improvements:
* Add build time check for all other platforms
* Add runtime check for all platforms
* Add various feature flags on/off

#### Issue #, if available

PR to introduce the error: https://github.com/aws-amplify/amplify-codegen/pull/489

PR to fix the error: https://github.com/aws-amplify/amplify-codegen/pull/500

#### Description of how you validated changes

Ran tests in circleci: https://app.circleci.com/pipelines/github/aws-amplify/amplify-codegen/2220/workflows/02eb292e-5b68-45f3-a56d-0cd9b2424c98

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-codegen/blob/main/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-codegen-e2e-core/src/init/cra.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    const s = { ...defaultSettings, ...settings };
+    const chain = spawn('npm', ['install'], { cwd, stripColors: true, disableCIDetection: s.disableCIDetection });
+
+    chain.run((err: Error) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve();
+      }
+    });
```
Comment (alharris-at) [inline packages/amplify-codegen-e2e-core/src/init/cra.ts:19]: `runAsync` is a nice way to reduce the if/else here. This block could be rewritten.

```typescript
return spawn('npm', ['install'], { cwd, stripColors: true, disableCIDetection: s.disableCIDetection }).runAsync();
```

Comment (alharris-at) [reply packages/amplify-codegen-e2e-core/src/init/cra.ts:19]: And same for the following fn as well.

File: packages/amplify-codegen-e2e-tests/test-apps/ts/scripts/set-schema.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+const fs = require('fs');
+const { schemas } = require('@aws-amplify/graphql-schema-test-library');
+
+const schemaName = process.argv.slice(2, 3)[0];
+if (!schemaName) {
+  throw new Error('A schema name must be supplied.');
+}
+const schema = schemas[schemaName];
+if (!schema) {
+  throw new Error(`${schemaName} does not exist in the schema test library.`);
```
Comment (alharris-at) [inline packages/amplify-codegen-e2e-tests/test-apps/ts/scripts/set-schema.js:10]: Maybe list possible options as well?

**Review by alharris-at — COMMENTED** [review]
Couple of comments, none of which are blocking.

The only question was whether moving the test app itself out of the `packages` directory resolves the need to have all the lint rules updated, setting it private so we don't accidentally publish, etc.

**Comment by dpilch** [comment]
Lint ignore would still be necessary because the lint script is [targeting all JS/TS files](https://github.com/aws-amplify/amplify-codegen/blob/main/package.json#L9) (we could update this to be more precise though). I think the test app is safe where it is because `amplify-codegen-e2e-tests` itself is also private. I don't really feel strongly on where it is though. I put it here because I though `../../` would be confusing/ugly in the test.

File: packages/amplify-codegen-e2e-core/src/init/cra.ts (hunk 1 above)
Comment (dpilch) [reply packages/amplify-codegen-e2e-core/src/init/cra.ts:19]: Doesn't look like we have a `runAsync` https://github.com/aws-amplify/amplify-codegen/blob/main/packages/amplify-codegen-e2e-core/src/utils/nexpect.ts#L56-L72.

File: .circleci/config.yml [hunk 3]
```diff
@@ -379,6 +379,14 @@ jobs:
     environment:
       TEST_SUITE: src/__tests__/graphql-codegen-ios.test.ts
       CLI_REGION: us-west-1
+  build-app-ts-e2e-test:
```
Comment (phani-srikar) [inline .circleci/config.yml:382]: These changes to config should go into `.circleci/config.base.yml` and I believe this file gets overridden by config generated by running `yarn split-e2e-tests`.

File: packages/amplify-codegen-e2e-tests/test-apps/ts/public/manifest.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  "short_name": "React App",
+  "name": "Create React App Sample",
+  "icons": [
+    {
+      "src": "favicon.ico",
+      "sizes": "64x64 32x32 24x24 16x16",
+      "type": "image/x-icon"
+    },
+    {
+      "src": "logo192.png",
```
Comment (phani-srikar) [inline packages/amplify-codegen-e2e-tests/test-apps/ts/public/manifest.json:11]: do you really need these images in your app?

File: packages/amplify-codegen-e2e-tests/test-apps/ts/src/setupTests.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+// jest-dom adds custom jest matchers for asserting on DOM nodes.
+// allows you to do things like:
+// expect(element).toHaveTextContent(/react/i)
+// learn more: https://github.com/testing-library/jest-dom
+import '@testing-library/jest-dom';
```
Comment (phani-srikar) [inline packages/amplify-codegen-e2e-tests/test-apps/ts/src/setupTests.ts:5]: blank file?

**Review by phani-srikar — CHANGES_REQUESTED** [review]
Looks good, can you take another pass and remove any boiler plate files from react app? Thanks.

File: packages/amplify-codegen-e2e-tests/test-apps/ts/public/manifest.json (hunk 4 above)
Comment (dpilch) [reply packages/amplify-codegen-e2e-tests/test-apps/ts/public/manifest.json]: Removed

File: packages/amplify-codegen-e2e-tests/test-apps/ts/src/setupTests.ts (hunk 5 above)
Comment (dpilch) [reply packages/amplify-codegen-e2e-tests/test-apps/ts/src/setupTests.ts]: Removed

File: packages/amplify-codegen-e2e-tests/test-apps/ts/package.json [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  "private": true,
+  "dependencies": {
+    "@testing-library/jest-dom": "^5.14.1",
+    "@testing-library/react": "^13.0.0",
+    "@testing-library/user-event": "^13.2.1",
+    "@types/jest": "^27.0.1",
+    "@types/node": "^16.7.13",
+    "@types/react": "^18.0.0",
+    "@types/react-dom": "^18.0.0",
+    "aws-amplify": "^4.3.43",
```
Comment (AaronZyLee) [inline packages/amplify-codegen-e2e-tests/test-apps/ts/package.json:13]: TODO reminder: we need to bump the version if there is a major version bump for `aws-amplify`

File: packages/amplify-codegen-e2e-tests/test-apps/ts/.graphqlconfig.yml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+projects:
```
Comment (AaronZyLee) [inline packages/amplify-codegen-e2e-tests/test-apps/ts/.graphqlconfig.yml:1]: This file can be removed if doc/type gen is not in scope of this test.

File: packages/amplify-codegen-e2e-tests/test-apps/ts/scripts/set-schema.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+const fs = require('fs');
```
Comment (AaronZyLee) [inline packages/amplify-codegen-e2e-tests/test-apps/ts/scripts/set-schema.js:1]: Is this script in use? I see there is a method `updateApiSchemaWithText` in e2e test used for setting up the schema.

**Review by AaronZyLee — CHANGES_REQUESTED** [review]

File: packages/amplify-codegen-e2e-tests/test-apps/ts/scripts/set-schema.js (hunk 8 above)
Comment (dpilch) [reply packages/amplify-codegen-e2e-tests/test-apps/ts/scripts/set-schema.js:1]: This is used for local testing outside of the test suite. It is not strictly necessary, but I found it useful.

File: packages/amplify-codegen-e2e-tests/test-apps/ts/package.json (hunk 6 above)
Comment (dpilch) [reply packages/amplify-codegen-e2e-tests/test-apps/ts/package.json:13]: This should already be v5 actually. Good catch

Comment (dpilch) [reply packages/amplify-codegen-e2e-tests/test-apps/ts/package.json:13]: Changed to `latest` so that it is always up to date.

**Review by AaronZyLee — APPROVED** [review]

**Review by alharris-at — APPROVED** [review]

**Pull request merged**